libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+                case FTP_OP::Rename:
+                default:
+                    // this was bad data, just nack it
+                    gcs().send_text(MAV_SEVERITY_DEBUG, "Unsupported FTP: %d", static_cast<int>(request.opcode));
+                    ftp_error(reply, FTP_ERROR::Fail);
+                    break;
+            }
+        }
+
+        // send our response backout to the system
+        while (!ftp.replies->push(reply)) { // we must fit the response, keep shoving it in
+            hal.scheduler->delay(10);
+        }
+        continue;
+    }
+}
+
+// calculates how much string length is needed to fit this in a list response
+int GCS_MAVLINK::emit_dir_entry(char *dest, size_t space, const char *path, const struct dirent * entry) {
+    const bool is_file = entry->d_type == DT_REG;
+
+    if (!is_file && entry->d_type != DT_DIR) {
+        return -1; // this just forces it so we can't send this back, it's easier then sending skips to a GCS
+    }
+
+    if (is_file) {
+        const size_t full_path_len = strlen(path) + strnlen(entry->d_name, 256); // FIXME: Really should do better then just hardcoding 256
+        char full_path[full_path_len + 2];
+        hal.util->snprintf(full_path, sizeof(full_path), "%s/%s", path, entry->d_name);
+        struct stat st;
+        if (AP::FS().stat(full_path, &st)) {
+            return -1;
+        }
+        return hal.util->snprintf(dest, space, "F%s\t%d\0", entry->d_name, (int)st.st_size);
+    } else {
+        return hal.util->snprintf(dest, space, "D%s\0", entry->d_name);
+    }
+}
+
+// list the contents of a directory, skip the offset number of entries before providing data
+void GCS_MAVLINK::ftp_list_dir(struct pending_ftp &request, struct pending_ftp &response) {
+    response.offset = request.offset; // this should be set for any failure condition for debugging
+
+    const size_t directory_name_size = strnlen((char *)request.data, sizeof(request.data));
+    // sanity check that our the request looks well formed
+    if ((directory_name_size != request.size) || (request.size == 0)) {
+        ftp_error(response, FTP_ERROR::InvalidDataSize);
+        return;
+    }
+
+    request.data[sizeof(request.data) - 1] = 0; // ensure the path is null terminated
+
+    // open the file
+    DIR *dir = AP::FS().opendir((char *)request.data);
+    if (dir == nullptr) {
+        ftp_error(response, FTP_ERROR::FailErrno);
+        AP::FS().closedir(dir);
+        return;
+    }
+
+    // burn the entries we don't care about
+    while (request.offset > 0) {
+        const struct dirent *entry = AP::FS().readdir(dir);
+        if(entry == nullptr) {
+            ftp_error(response, FTP_ERROR::EOF_);
+            AP::FS().closedir(dir);
+            return;
+        }
+
+        // check how much space would be needed to emit the listing
+        const int needed_space = emit_dir_entry((char *)response.data, sizeof(request.data), (char *)request.data, entry);
```

Comment:
This function really needs a better name.

Reply:
Suggestions? I find the name to be quite logical, its role in life is to emit a properly formatted dir entry into the buffer.

Reply:
I would go with gen_dirlist_entry, emit_dir_entry has a connotation that its actually sending something which it is not.

Reply:
fixed